packages/node/src/indexer/api.service.ts

```diff
@@ -62,52 +163,164 @@ export class ApiService implements OnApplicationShutdown {
-          this.networkMeta.genesisHash
-        }). Please check that the RPC endpoint is actually for your desired network or update the genesisHash.`,
-      );
-      logger.error(err, err.message);
-      throw err;
+        if (
+          network.chainId &&
+          network.chainId !== this.networkMeta.genesisHash
+        ) {
+          const err = new Error(
+            `Network chainId doesn't match expected genesisHash. Your SubQuery project is expecting to index data from "${
+              network.chainId ?? network.genesisHash
+            }", however the endpoint that you are connecting to is different("${
+              this.networkMeta.genesisHash
+            }). Please check that the RPC endpoint is actually for your desired network or update the genesisHash.`,
+          );
+          logger.error(err, err.message);
+          throw err;
+        }
+      } else {
+        const chain = api.runtimeChain.toString();
+        if (this.networkMeta.chain !== chain) {
+          throw this.metadataMismatchError(
+            'Runtime Chain',
+            this.networkMeta.chain,
+            chain,
+          );
+        }
+
+        const specName = api.runtimeVersion.specName.toString();
+        if (this.networkMeta.specName !== specName) {
+          throw this.metadataMismatchError(
+            'Spec Name',
+            this.networkMeta.specName,
+            specName,
+          );
+        }
+
+        const genesisHash = api.genesisHash.toString();
+        if (this.networkMeta.genesisHash !== genesisHash) {
+          throw this.metadataMismatchError(
+            'Genesis Hash',
+            this.networkMeta.genesisHash,
+            genesisHash,
+          );
+        }
+      }
+
+      this.connectionPool[i] = api;
+      this.apiOptions.push(apiOption);
+      this.endpoints.push(endpoint);
     }
 
+    this.loadBalancer = new ApiLoadBalancer(
+      this.numConnections,
+      this.nodeConfig.batchSize,
+    );
+
     return this;
   }
 
-  getApi(): ApiPromise {
-    return this.api;
+  async attemptReconnects(): Promise<void> {
+    if (Object.keys(this.disconnectedApis).length === 0) {
+      return;
+    }
+    await Promise.all(
+      Object.keys(this.disconnectedApis).map(async (key) => {
+        try {
+          logger.info(`Attempting to reconnect to ${this.endpoints[key]}`);
+          await this.disconnectedApis[key].connect();
+          logger.info(`Reconnected to ${this.endpoints[key]}`);
+        } catch (e) {
+          logger.error(e);
+          logger.error(`Failed to reconnect to ${this.endpoints[key]}`);
+        }
+      }),
+    );
+  }
+
+  async addToDisconnectedApiIndices(index: number): Promise<void> {
+    if (this.disconnectedApiIndices.includes(index)) {
+      return;
+    }
+    await this.connectionPool[index].disconnect();
+    this.disconnectedApiIndices.push(index);
+  }
+
+  getApi(index?: number): ApiPromise {
```

Comment:
```suggestion
  get api(): ApiPromise {
```

Outside of this class there should be no concept of more than one api instance